app/services/idempotency/store.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+require 'redis'
+
+module Idempotency
+  class Store
+    # NB the caching vs conflict logic is:
+    # *  re-using the idempotency key with the same request for 10 mins will return the same cached response
+    # *  re-using the idempotency key with the same request after 10 mins to 12 hours will raise a conflict error
+    # *  re-using the idempotency key with a different request will raise a conflict error, for 12 hours
+
+    CACHE_RESPONSE_TTL = 600 # 10 minutes
+    CONFLICT_TTL = 43_200 # 12 hours
+
+    attr_reader :idempotency_key, :conflict_key, :cache_response_key
+
+    def initialize(request)
+      # NB: we need to do a case-insensitive match for IDEMPOTENCY_KEY; rails does not do this automatically if the key contains an underscore
+      @idempotency_key = request.headers.find { |key, _v| key =~ /\AIDEMPOTENCY[\_\-]KEY\Z/i }&.last
+      if @idempotency_key.present?
+        @conflict_key = "conf|#{idempotency_key}"
+        @cache_response_key = "resp|#{idempotency_key}|#{request_hash(request)}"
+      end
+    end
+
+    def get
+      return if @idempotency_key.blank?
+
+      # Return the cached response if it matches the idempotency key and request
+      cached_response = redis.hgetall(cache_response_key)
+
+      return cached_response if cached_response.present?
+
+      # Otherwise, raise a conflict error of the idempotency key exists, or return nil if not
+      conflict = redis.get(conflict_key)
+      raise ConflictError, idempotency_key unless conflict.nil?
+    end
+
+    def set(response)
+      return if @idempotency_key.blank?
+
+      # cache the response for a short time
+      redis.hmset(cache_response_key, *response)
```

Comment:
Maybe we could leverage [pipelining](https://github.com/redis/redis-rb#pipelining) here?

Reply:
I'm no longer using lowlevel redis access - instead the rails.cache object will use the `hiredis` gem for high-performance redis